File: app/src/main/java/br/com/dbccompany/sicred/presentation/board/ui/BoardViewModel.java
package br.com.dbccompany.sicred.presentation.board.ui;

import android.provider.CalendarContract;
import android.util.Log;

import androidx.hilt.Assisted;
import androidx.hilt.lifecycle.ViewModelInject;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.SavedStateHandle;
import androidx.lifecycle.ViewModel;

import java.util.List;

import br.com.dbccompany.sicred.data.datasource.remote.event.EventRemoteDataSource;
import br.com.dbccompany.sicred.data.repositories.EventRepository;
import br.com.dbccompany.sicred.domain.model.Event;
import br.com.dbccompany.sicred.domain.usecase.ListEventsUseCase;
import io.reactivex.Emitter;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.annotations.NonNull;
import io.reactivex.observers.DisposableSingleObserver;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class BoardViewModel extends ViewModel {
    private final EventRepository eventRepository;
    private final SavedStateHandle savedStateHandle;
    private final ListEventsUseCase listEventsUseCase;
    private MutableLiveData<List<Event>> eventsLiveData = new MutableLiveData<>();
    private MutableLiveData<Boolean> showError = new MutableLiveData<>();
    private MutableLiveData<Boolean> showProgress = new MutableLiveData<>();
    private MutableLiveData<Boolean> showContent = new MutableLiveData<>();

    @ViewModelInject
    public BoardViewModel(EventRepository eventRepository,
                          @Assisted SavedStateHandle savedStateHandle,
                          ListEventsUseCase listEventsUseCase
    ) {
        this.eventRepository = eventRepository;
        this.savedStateHandle = savedStateHandle;
        this.listEventsUseCase = listEventsUseCase;
        loadEvents();
    }

    private static final String TAG = "BoardViewModel";
    public void loadEvents() {
        setLoadState();
        handleResult(listEventsUseCase.invoke());
        Log.d(TAG, "loadEvents: ");
    }
    public void setLoadState(){
        showProgress.postValue(true);
        showError.postValue(false);
        showContent.postValue(false);
    }
    public void handleResult(Call<List<Event>> result){
         eventRepository.getEvents().enqueue(new Callback<List<Event>>() {
            @Override
            public void onResponse(Call<List<Event>> call, Response<List<Event>> response) {
                setContentState(response.body());
            }

            @Override
            public void onFailure(Call<List<Event>> call, Throwable t) {
                setErrorState();
            }
        });
    }


    public void setContentState(List<Event> eventList){
        showContent.postValue(true);
        eventsLiveData.postValue(eventList);
        showProgress.postValue(false);
        showError.postValue(false);
    }

    public void setErrorState(){
        showError.postValue(true);
        showProgress.postValue(false);
        showContent.postValue(false);
    }

    public EventRepository getEventRepository() {
        return eventRepository;
    }

    public SavedStateHandle getSavedStateHandle() {
        return savedStateHandle;
    }

    public ListEventsUseCase getListEventsUseCase() {
        return listEventsUseCase;
    }

    public MutableLiveData<List<Event>> getEventsLiveData() {
        return eventsLiveData;
    }

    public void setEventsLiveData(MutableLiveData<List<Event>> eventsLiveData) {
        this.eventsLiveData = eventsLiveData;
    }

    public MutableLiveData<Boolean> getShowError() {
        return showError;
    }

    public void setShowError(MutableLiveData<Boolean> showError) {
        this.showError = showError;
    }

    public MutableLiveData<Boolean> getShowProgress() {
        return showProgress;
    }

    public void setShowProgress(MutableLiveData<Boolean> showProgress) {
        this.showProgress = showProgress;
    }

    public MutableLiveData<Boolean> getShowContent() {
        return showContent;
    }

    public void setShowContent(MutableLiveData<Boolean> showContent) {
        this.showContent = showContent;
    }
}

File: app/src/main/java/br/com/diogomoreno/events/presentation/board/ui/BoardViewModel.kt
package br.com.diogomoreno.events.presentation.board.ui

import android.util.Log
import androidx.hilt.Assisted
import androidx.hilt.lifecycle.ViewModelInject
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import br.com.diogomoreno.events.data.repositories.EventRepository
import br.com.diogomoreno.events.domain.model.Event
import br.com.diogomoreno.events.domain.usecase.ListEventsUseCase
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class BoardViewModel @ViewModelInject constructor(val eventRepository: EventRepository,
                                                  @param:Assisted val savedStateHandle: SavedStateHandle,
                                                  val listEventsUseCase: ListEventsUseCase
) : ViewModel() {
    var eventsLiveData = MutableLiveData<MutableList<Event?>?>()
    var showError = MutableLiveData<Boolean>()
    var showProgress = MutableLiveData<Boolean?>()
    var showContent = MutableLiveData<Boolean>()
    fun loadEvents() {
        setLoadState()
        handleResult(listEventsUseCase.invoke())
        Log.d(TAG, "loadEvents: ")
    }

    fun setLoadState() {
        showProgress.postValue(true)
        showError.postValue(false)
        showContent.postValue(false)
    }

    fun handleResult(result: Call<MutableList<Event?>?>) {
        eventRepository.events?.enqueue(object : Callback<MutableList<Event?>?> {
            override fun onFailure(call: Call<MutableList<Event?>?>, t: Throwable) {
                setErrorState()
            }

            override fun onResponse(call: Call<MutableList<Event?>?>, response: Response<MutableList<Event?>?>) {
                setContentState(response.body())
            }


        })
    }

    fun setContentState(eventList: MutableList<Event?>?) {
        showContent.postValue(true)
        eventsLiveData.postValue(eventList)
        showProgress.postValue(false)
        showError.postValue(false)
    }

    fun setErrorState() {
        showError.postValue(true)
        showProgress.postValue(false)
        showContent.postValue(false)
    }

    companion object {
        private const val TAG = "BoardViewModel"
    }

    init {
        loadEvents()
    }
}